### PR TITLE
dbt 1.3.0 upgrade

### LINKED ISSUE #151: upgrade to support dbt-core v1.3.0
## Background

The latest release cut for 1.3.0, `dbt-core==1.3.0rc2` was published on October 3, 2022 ([PyPI](https://pypi.org/project/dbt-core/1.3.0rc2/) | [Github](https://github.com/dbt-labs/dbt-core/releases/tag/v1.3.0rc2)). We are targeting releasing the official cut of 1.3.0 in time for the week of October 16 (in time for Coalesce conference).

We're trying to establish a following precedent w.r.t. minor versions:
Partner adapter maintainers release their adapter's minor version within four weeks of the initial RC being released. Given the delay on our side in notifying you, we'd like to set a target date of November 7 (four weeks from today) for maintainers to release their minor version

| Timeframe   | Date (intended) | Date (Actual) | Event                                                  |
| ----------- | --------------- | ------------- | ------------------------------------------------------ |
| D - 3 weeks | Sep 21          | Oct 10        | dbt Labs informs maintainers of upcoming minor release |
| D - 2 weeks | Sep 28          | Sep 28        | core 1.3 RC is released                                |
| Day D       | October 12      | Oct 12        | core 1.3 official is published                         |
| D + 2 weeks | October 26      | Nov 7         | dbt-adapter 1.3 is published                           |

## How to upgrade

https://github.com/dbt-labs/dbt-core/discussions/6011 is an open discussion with more detailed information, and https://github.com/dbt-labs/dbt-core/issues/6040 is for keeping track of the community's progress on releasing 1.2.0

Below is a checklist of work that would enable a successful 1.2.0 release of your adapter.

- [ ] Python Models (if applicable)
- [ ] Incremental Materialization: cleanup and standardization
- [ ] More functional adapter tests to inherit

### PR DESCRIPTION
**Pull request by mdesmet**
## Overview

Resolves #151

## Checklist

- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [ ] `README.md` updated and added information about my change
- [ ] I have run `changie new` to [create a changelog entry](https://github.com/starburstdata/dbt-trino/blob/master/CONTRIBUTING.md#Adding-CHANGELOG-Entry)

### CODE REVIEW

File: dbt/include/trino/macros/materializations/incremental.sql [hunk 1]
```diff
@@ -1,37 +1,106 @@
...
-    {% do return(get_delete_insert_merge_sql(target_relation, tmp_relation, unique_key, dest_columns)) %}
-  {% elif strategy == 'merge' %}
-    {% do return(get_merge_sql(target_relation, tmp_relation, unique_key, dest_columns, predicates=none)) %}
+{% materialization incremental, adapter='trino', supported_languages=['sql'] -%}
+
+  {#-- Set vars --#}
+  {%- set full_refresh_mode = (should_full_refresh()) -%}
+  {%- set language = model['language'] -%}
+  {#-- TODO: figure out why we need to add table explicitly, otherwise persist_docs doesn't know the type --#}
+  {% set target_relation = this.incorporate(type='table') %}
```
Comment (mdesmet) [inline dbt/include/trino/macros/materializations/incremental.sql:26]: Interestingly the snowflake adapter doesn't have to add `.incorporate(type='table')`, might be a bug in the relation implementation.

File: dbt/include/trino/macros/materializations/incremental.sql [hunk 2]
```diff
@@ -1,37 +1,106 @@
...
+       If we are running multiple statements (DELETE + INSERT),
+       we must first save the model query results as a temporary table
+       in order to guarantee consistent inputs to both statements.
+
+       If we are running a single statement (MERGE or INSERT alone),
+       we can save the model query definition as a view instead,
+       for faster overall incremental processing.
+  #} */
+
+  {% if language == 'sql' and (strategy in ('default', 'append', 'merge') or (unique_key is none)) %}
```
Comment (mdesmet) [inline dbt/include/trino/macros/materializations/incremental.sql:13]: I suggest to add another config something like `disable_views` to enable using connectors that don't support views

File: dbt/include/trino/macros/utils/timestamps.sql [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{% macro trino__current_timestamp() -%}
+    current_timestamp
+{%- endmacro %}
+
+{% macro trino__snapshot_string_as_time(timestamp) %}
+    {%- set result = "TIMESTAMP '" ~ timestamp ~ "'" -%}
```
Comment (mdesmet) [inline dbt/include/trino/macros/utils/timestamps.sql:6]: probably should be lowercase as well to keep it consistent.

File: tests/functional/adapter/test_grants.py [hunk 4]
```diff
@@ -30,3 +31,17 @@ def assert_expected_grants_match_actual(self, project, relation_name, expected_g
...
         diff_b = BaseContext.diff_of_two_dicts(expected_grants, actual_grants)
         assert diff_a == diff_b == {}
+
+
+@pytest.mark.hive
+# TODO: setup Galaxy and Starbust tests
+#   See https://github.com/starburstdata/dbt-trino/issues/147
+#   and also https://github.com/starburstdata/dbt-trino/issues/146
+@pytest.mark.skip_profile("starburst_galaxy")
+@pytest.mark.skip(reason="Hive doesn't raise errors on invalid roles")
```
Comment (mdesmet) [inline tests/functional/adapter/test_grants.py:41]: To be investigated. Seems like Hive doesn't raise an error on invalid grants.

Also these exceptions might be different for Galaxy, SEP or HIve.

File: tests/functional/adapter/utils/test_utils.py [hunk 5]
```diff
@@ -33,6 +37,22 @@ class TestAnyValue(BaseAnyValue):
     pass
 
 
+# Not working because of https://github.com/trinodb/trino/issues/13
+@pytest.mark.skip(reason="No wqy to concat an array with null or empty array")
```
Comment (mdesmet) [inline tests/functional/adapter/utils/test_utils.py:41]: This is unfortunate. I have tried multiple workarounds but the Trino parser will always fail on the `concat(null, ARRAY[1,2])` even if you put it behind if, try or case when.

File: dbt/adapters/trino/column.py [hunk 6]
```diff
@@ -13,6 +13,39 @@ class TrinoColumn(Column):
         "INTEGER": "INT",
     }
```
Comment (mdesmet) [inline dbt/adapters/trino/column.py:15]: I still need to debug the following changes. If i remember correctly this might be related to type detection from agate but not entirely sure.

File: .changes/1.3.0.md [hunk 7]
```diff
@@ -1,2 +1 @@
 ## dbt-trino 1.3.0 - October 13, 2022
-
```
Comment (hovaesco) [inline .changes/1.3.0.md:2]: unrelated

File: tests/functional/adapter/utils/test_utils.py [hunk 8]
```diff
@@ -33,6 +37,22 @@ class TestAnyValue(BaseAnyValue):
...
+class TestArrayAppend(BaseArrayAppend):
+    pass
+
+
+# Not working because of https://github.com/trinodb/trino/issues/13
+@pytest.mark.skip(reason="No wqy to concat an array with null or empty array")
```
Comment (hovaesco) [inline tests/functional/adapter/utils/test_utils.py:47]: a typo in way

File: tests/functional/adapter/utils/test_utils.py (hunk 5 above)
Comment (hovaesco) [reply tests/functional/adapter/utils/test_utils.py:41]: a typo in way

File: tests/functional/adapter/utils/test_utils.py [hunk 9]
```diff
@@ -45,6 +65,11 @@ class TestConcat(BaseConcat):
     pass
 
 
+# Use either BaseCurrentTimestampAware or BaseCurrentTimestampNaive but not both
```
Comment (hovaesco) [inline tests/functional/adapter/utils/test_utils.py:68]: some more explanation would be nice to have here since it's no clear why

File: tests/functional/adapter/utils/test_data_types.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    pass
+
+
+class TestTypeInt(BaseTypeInt):
+    pass
+
+
+class TestTypeNumeric(BaseTypeNumeric):
+    def numeric_fixture_type(self):
+        return "decimal(28,6)"
```
Comment (hovaesco) [inline tests/functional/adapter/utils/test_data_types.py:24]: Should precision be 38?

File: tests/functional/adapter/test_grants.py (hunk 4 above)
Comment (hovaesco) [reply tests/functional/adapter/test_grants.py:41]: Is it now skipped for Trino as well, if yes then `@pytest.mark.skip_profile("starburst_galaxy")` is obsolete.

File: tests/functional/adapter/materialization/test_incremental_merge.py [hunk 11]
```diff
@@ -135,6 +135,8 @@ def project_config_update(self):
 
 
 @pytest.mark.delta
+# TODO: re-enable when https://github.com/trinodb/trino/pull/11763 is merged
```
Comment (hovaesco) [inline tests/functional/adapter/materialization/test_incremental_merge.py:138]: Please add an issue to enable these tests in dbt-trino.

File: dbt/include/trino/macros/utils/timestamps.sql [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+    {{ return(result) }}
+{% endmacro %}
+
+{% macro trino__current_timestamp_backcompat() -%}
```
Comment (hovaesco) [inline dbt/include/trino/macros/utils/timestamps.sql:10]: why it's needed looks like a copy of `trino__current_timestamp_backcompat`

File: dbt/include/trino/macros/utils/timestamps.sql (hunk 3 above)
Comment (hovaesco) [reply dbt/include/trino/macros/utils/timestamps.sql:6]: yeah let's keep it consistent

File: dbt/adapters/trino/column.py (hunk 6 above)
Comment (hovaesco) [reply dbt/adapters/trino/column.py:15]: Please explain why these changes are required

File: dbt/include/trino/macros/utils/timestamps.sql [hunk 13]
```diff
@@ -0,0 +1,12 @@
+{% macro trino__current_timestamp() -%}
```
Comment (hovaesco) [inline dbt/include/trino/macros/utils/timestamps.sql:1]: Why plural name of the file `timestamps.sql`?

**Review by hovaesco — COMMENTED** [review]
Lots of changes in `dbt/adapters/trino/`, I'm not sure whether are of these are required. Let's keep only dbt 1.3 related changes here and extract separate to other PR.

File: tests/functional/adapter/utils/test_utils.py (hunk 9 above)
Comment (mdesmet) [reply tests/functional/adapter/utils/test_utils.py:68]: I think before dbt didn't distinguish between a time zone aware timestamp and a naive timestamp while we always worked with timezone aware (UTC) timestamps.

File: .changes/1.3.0.md (hunk 7 above)
Comment (mdesmet) [reply .changes/1.3.0.md:2]: Actually this is related to the version bump. this is done by pre-commit check.

File: dbt/include/trino/macros/utils/timestamps.sql (hunk 13 above)
Comment (mdesmet) [reply dbt/include/trino/macros/utils/timestamps.sql:1]: I have tried to mimic as much the official dbt plugins so we can easily follow course in the next upgrades.

**Comment by hovaesco** [comment]
Please create the PR directly from the repo not from fork to trigger Galaxy tests.

**Comment by mdesmet** [comment]
Replaced by #161
